### PR TITLE
Whisper v6 libp2p v2

### PR DESCRIPTION
**Pull request by gballet**
Rewrite after the feedback I got from https://github.com/ethereum/go-ethereum/pull/15979.

With this PR, `wnode` can now send messages to other `wnode` instances. The bootnode address have the form `/ip4/127.0.0.1/tcp/3001/ipfs/Qma5cHmMTpzxTiLJpwQ3JM7vn8viZcEtKfttWQAyKYDizm`, where `Qma5cHmMTpzxTiLJpwQ3JM7vn8viZcEtKfttWQAyKYDizm` is an IPFS peer id.

Currently missing from this PR, which will be added in the coming days:
  * [x] @fjl suggested not create my own protocol for the exchange of packages, and instead to wrap RLP + Bufios. Implemented in #16430
  * [ ] The list of peers is currently exchanged through the `handshake()` function, I still have to check that this is enough to create a fully connected network of Whisper-enabled peers. My hunch is that it is not.
  * [ ] @fjl also suggested to use the protocol multiplexer to have Whisper as one of the many supported protocols. This is an important step in offering a fully compliant implementation of `geth` on libp2p.
  * [x] Enable `secio`
  * [x] Given the time it took me to create this PR, the `master` branch has drifted, merge it
  * [x] Add libp2p to `vendor` see https://github.com/ethereum/go-ethereum/pull/16422
  * [x] add mutexes to remove data races in libp2p peer management code

### CODE REVIEW

File: cmd/wnode/main.go [hunk 1]
```diff
@@ -189,15 +195,39 @@ func initialize() {
...
 
 	if *bootstrapMode {
-		if len(*argIP) == 0 {
-			argIP = scanLineA("Please enter your IP and port (e.g. 127.0.0.1:30348): ")
+		if !*useLibP2P {
+			if len(*argIP) == 0 {
+				argIP = scanLineA("Please enter your IP and port (e.g. 127.0.0.1:30348): ")
+			}
+		} else {
+			if *argPort == 0 {
```
Comment (gluk256) [inline cmd/wnode/main.go]: please change to:
} else if *argPort == 0 {

**Comment by raulk** [comment]
Hey @gballet, this is Raúl from the libp2p team at Protocol Labs. Wondered if you needed anything from our end, particularly with bullets 2 and 3 in the PR description?

**Comment by gballet** [comment]
@raulk thanks; this PR is going to be overhauled shortly to 1) account for some changes in the P2P layer on our side; and 2) expand beyond Whisper (point number 3). What you can do to help is:

  * Review this PR and tell us if I missed something big
  * Check if point number 2 (the first unchecked one) is correctly implemented

Thanks!

**Comment by raulk** [comment]
@gballet 👍 Please give me a shoutout when it's time for that.

**Comment by raulk** [comment]
Using [`blankhost`](https://github.com/libp2p/go-libp2p-blankhost) instead of `basichost` should give us more fine-grained control on the vendored packages. `basichost` imports circuit relays, mdns, and a few other things I suspect are not needed here.

**Comment by gballet** [comment]
@mgoelzer

**Comment by holiman** [comment]
What's the status here @gballet ? Wip? Abandoned? still want it?

**Comment by gballet** [comment]
it's meant to serve as a reference, although this PR proper won't be merged.

**Comment by gballet** [comment]
Won't be merged :disappointed:

**Pull request closed without merging**